### ISSUE TITLE
Contextual type isn't provided right when generic type discriminator exists in two properties

### ISSUE THREAD

**Issue by osasson** [microsoft/TypeScript#57548]
### 🔎 Search Terms

"generic type discriminator"

### 🕗 Version & Regression Information

Version 5.2.2

### 💻 Code

```ts
type A = {
	1: { dt: string },
	2: { num: number },
};
type B = { [K in keyof A]: {
	extra: { id: number, reason: K }
	data: A[K]
 } }[keyof A];

 type C = { [K in keyof A]: {
	data: A[K] & { reason: K }
 } }[keyof A];

const test:B = {} as any;
if (test.extra.reason === 2) {
	// Property 'num' does not exist on type '{ dt: string; } | { num: number; }'. Property 'num' does not exist on type '{ dt: string; }'.ts(2339)
	console.log(test.data.num);
}
// expected no error
const test1:C = {} as any;
if (test1.data.reason == 2) {
	console.log(test1.data.num);
}

// expected error
const test2:B = {
	extra: { id: 1, reason: 2 },
	data: { dt: 'string' },
};

// expected no error
const test3:B = {
	extra: { id: 1, reason: 2 },
	data: { num: 123 },
};


```

### 🙁 Actual behavior

Typescript does not discriminate the type according to its constraints

### 🙂 Expected behavior

discrimination should occur

**Comment by RyanCavanaugh** [comment]
Duplicate #18758
